### PR TITLE
U/angon/capitalize

### PR DESCRIPTION
**Pull request by aniagonzalez**
This change adds a feature to change case of selection to UPPERCASE, lowercase or Capitalize Each Word options.

image

It is implemented using applyInlineStyle to traverse the elements in the selection. 
To get the capitalized test it uses text-transform which handles the language-specific mapping rules.

TO REVIEW:
- For correct capitalization to happen in all languages, we would need the editor to know of the language (through xml-lang with current implementation). See this example, the first test passes but the second fails:
image

- I am not sure the added SPAN elements are acceptable. Please see file changeCapitalizationTest.ts

- I grabbed an open source icon because we didn't have an appropriate one in our codebase. Should I change it to just a text button?

### CODE REVIEW

File: packages/roosterjs-editor-api/lib/format/changeCapitalization.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+            }
+        }
+    });
+
+    function preprocessText(textToPreprocess: string) {
+        const lowercaseText = getCapitalizedText(textToPreprocess, Capitalization.Lowercase);
+        return lowercaseText;
+    }
+
+    function getCapitalizedText(originalText: string, capitalization: Capitalization): string {
```
Comment (aniagonzalez) [inline packages/roosterjs-editor-api/lib/format/changeCapitalization.ts:28]: Per our discussion, I will be changing this method to accept an optional LANG.

File: packages/roosterjs-editor-api/lib/format/changeCapitalization.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import { Capitalization, NodeType } from 'roosterjs-editor-types';
+import applyInlineStyle from '../utils/applyInlineStyle';
+import { getFirstLeafNode, getNextLeafSibling } from 'roosterjs-editor-dom';
+
+/**
+ * Change the capitalization of text in the selection
+ * @param editor The editor instance
+ * @param capitalization The case option
+ */
+export default function g(editor: IEditor, capitalization: Capitalization, language?: string) {
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/changeCapitalization.ts:11]: Use a meaningful function name

File: packages/roosterjs-editor-api/lib/format/changeCapitalization.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        //     tempNode.style.textTransform = capitalization;
+        //     transformedText = tempNode.innerText;
+
+        //     document.body.removeChild(tempNode);
+        // }
+        // return transformedText;
+
+        switch (capitalization) {
+            case Capitalization.Lowercase:
+                return originalText.toLocaleLowerCase(language);
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/changeCapitalization.ts:44]: If language is undefined, will it respect browser's language setting?

File: packages/roosterjs-editor-api/lib/format/changeCapitalization.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        //     if (language) {
+        //         tempNode.lang = language;
+        //     }
+        //     tempNode.innerText = originalText;
+        //     tempNode.style.textTransform = capitalization;
+        //     transformedText = tempNode.innerText;
+
+        //     document.body.removeChild(tempNode);
+        // }
+        // return transformedText;
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/changeCapitalization.ts:40]: Remove unused code /comment

Comment (aniagonzalez) [reply packages/roosterjs-editor-api/lib/format/changeCapitalization.ts]: Sorry, I forgot I had already started this PR when I pushed my last changes!

File: packages/roosterjs-editor-api/lib/format/changeCapitalization.ts (hunk 3 above)
Comment (aniagonzalez) [reply packages/roosterjs-editor-api/lib/format/changeCapitalization.ts:27]: Yes, it is an optional parameter, the default locale is the host environment’s current locale.

File: packages/roosterjs-editor-api/lib/format/changeCapitalization.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import { IEditor } from 'roosterjs-editor-types';
+import { Capitalization, NodeType } from 'roosterjs-editor-types';
+import applyInlineStyle from '../utils/applyInlineStyle';
+import { getFirstLeafNode, getNextLeafSibling } from 'roosterjs-editor-dom';
+
+/**
+ * Change the capitalization of text in the selection
+ * @param editor The editor instance
+ * @param capitalization The case option
+ */
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/changeCapitalization.ts:10]: Add comment for language, and explain the default behavior

File: packages/roosterjs-editor-api/lib/format/changeCapitalization.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+                return wordArray.join(' ');
+            case Capitalization.Sentence:
+                // TODO: Add rules on punctuation for internationalization - TASK 104769
+                const punctuationMarks = '[\\.\\!\\?]';
+                // Find a match of a word character either:
+                // at the beginning of a string with or without preceding whitespace
+                // or preceded by a punctuation mark and at least one whitespace.
+                const regex = new RegExp('^\\s*\\w|' + punctuationMarks + '\\s+\\w', 'g');
+                return originalText.toLocaleLowerCase(language).replace(regex, match => {
+                    return match.toUpperCase();
```
Comment (JiuqingSong) [inline packages/roosterjs-editor-api/lib/format/changeCapitalization.ts:45]: toUpperCase or toLocaleUpperCase ?

Comment (aniagonzalez) [reply packages/roosterjs-editor-api/lib/format/changeCapitalization.ts:45]: Thanks for catching this!

**Review by JiuqingSong — APPROVED** [review]

**Pull request merged**